### PR TITLE
Use the git CLI with cargo

### PR DESCRIPTION
**Pull request by bmw**
Hopefully this makes things more stable. This is based on Alex's suggestion [here](https://opensource.eff.org/eff-open-source/pl/ouf996zuxjnkdxwq81bihxak7e). I'm testing building the snaps and docker images for all architectures with this change at https://dev.azure.com/certbot/certbot/_build/results?buildId=5227&view=results.

### CODE REVIEW

**Comment by wgreenberg** [comment]
this lgtm, though presumably we'll wanna wait for the azure tests you linked to pass before merging?

**Comment by bmw** [comment]
Yeah. I suspect it'll pass, but it's not a bad idea to wait and make sure.

**Comment by bmw** [comment]
The tests passed :tada:

File: tools/docker/core/Dockerfile [hunk 1]
```diff
@@ -26,6 +26,10 @@ RUN apk add --no-cache --virtual .certbot-deps \
         ca-certificates \
         binutils
 
+# We set this environment variable and install git while building to try and
+# increase the stability of fetching the rust crates needed to build the
+# cryptography library
+ARG CARGO_NET_GIT_FETCH_WITH_CLI=true
```
Comment (alexzorin) [inline tools/docker/core/Dockerfile:32]: Does `ARG` actually propagate to an environment variable?

My reading of https://docs.docker.com/engine/reference/builder/#arg is that `ARG` allows for parameterization within the scope of the Dockerfile, but you have to use `ENV` separately to make an environment variable from it.

Their example is:

```Dockerfile
ARG CONT_IMG_VER
ENV CONT_IMG_VER=${CONT_IMG_VER:-v1.0.0}
```

We're not parameterizing this for the snap, so maybe: 

```suggestion
ENV CARGO_NET_GIT_FETCH_WITH_CLI=true
```

**Review by alexzorin — COMMENTED** [review]
Another question I have is about the DNS plugin build process.

1. When the Docker build runs for plugins, does pip see the `cryptography` that was built in the base image? Or does it build it again? If the latter, we might need this environment variable there as well.
2. Ditto for the `snapcraft.yml` generated by `generate_dnsplugins_snapcraft.sh`. If one of our DNS plugins requires `cryptography` (via `lexicon`→`requests` perhaps), it's not getting it from the main snap, is it?

File: tools/docker/core/Dockerfile (hunk 1 above)
Comment (bmw) [reply tools/docker/core/Dockerfile:32]: I agree Docker's documentation isn't very clear there. These two sources are a bit clearer though:

* https://docs.docker.com/engine/reference/commandline/build/#set-build-time-variables---build-arg
* https://stackoverflow.com/questions/39597925/how-do-i-set-environment-variables-during-the-build-in-docker

My understanding is while you can override `ARG`, it sets an environment variable that's only available at build time which I think is what we want.

I also tested it and it seems to work:
```
cat << EOF > Dockerfile
FROM ubuntu
ARG FOO=bar
RUN env > ~/myenv
ENTRYPOINT cat ~/myenv
EOF
docker build -t foo .
docker run --rm foo
```
outputs
```
HOME=/root
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
FOO=bar
PWD=/
```

**Comment by bmw** [comment]
> 1. When the Docker build runs for plugins, does pip see the `cryptography` that was built in the base image? Or does it build it again? If the latter, we might need this environment variable there as well.

It has `cryptography` from the base image. [Our plugin's Dockerfile builds on the base image](https://github.com/certbot/certbot/blob/6e8f58e3f618c887df10c7640c3c148fc62cbf8c/tools/docker/plugin/Dockerfile#L3) and you can see this in effect in Azure places like [here](https://dev.azure.com/certbot/certbot/_build/results?buildId=5232&view=logs&j=db9d590d-ed27-5cbb-1855-2594371ebd55&t=9950ec7a-c9d6-5c77-fe7e-8b5a3934be17&l=405) where shortly after executing that command, the output is:
```
Requirement already satisfied: cryptography>=2.5.0 in /usr/local/lib/python3.8/site-packages (from acme>=1.25.0.dev0->certbot-dns-dnsmadeeasy==1.25.0.dev0) (36.0.1)
```

> 2. Ditto for the `snapcraft.yml` generated by `generate_dnsplugins_snapcraft.sh`. If one of our DNS plugins requires `cryptography` (via `lexicon`→`requests` perhaps), it's not getting it from the main snap, is it? 

Our snaps are not built from a common Certbot base image. Instead, if the plugin also requires `cryptography`, it'll include it's own version of it. I think this is taken care of though by me setting the variable in `tools/snap/generate_dnsplugins_snapcraft.sh` right?

File: tools/docker/core/Dockerfile (hunk 1 above)
Comment (alexzorin) [reply tools/docker/core/Dockerfile:32]: TIL, thanks.

**Comment by alexzorin** [comment]
Ah, I'm not sure how I missed the `generate_dnsplugins_snapcraft.sh` change in this PR. Sorry! LGTM.

**Review by alexzorin — APPROVED** [review]

**Pull request merged**